### PR TITLE
[SYCL] Use of Unified Shared Memory

### PR DESCRIPTION
**Pull request by Michoumichmich**
This PR changes the GPU backend to use USM instead of `sycl::buffer`s that can lead to implicit data movements. 
This allows also to use the same interface for the `c-sycl_intel` target. It also addresses the `FIXME` comment given that we can allocate "void" memory. 

Signed-off-by: Michel Migdal michel.migdal@codeplay.com

### CODE REVIEW

File: backends/c-sycl_intel/include/backend.hpp [hunk 1]
```diff
@@ -2,7 +2,7 @@
 // Licensed under the MIT License.
 
 //; eval_flags(c-sycl_intel): [dpcpp] -ldl -lpthread
-//; eval_flags(c-sycl_cuda): [/usr/local/dpcpp-cuda/bin/clang++] -ldl -I/usr/local/dpcpp-cuda/include/sycl -L/usr/local/dpcpp-cuda/lib -lsycl -fsycl -fsycl-targets=nvptx64-nvidia-cuda-sycldevice -fsycl-unnamed-lambda -lpthread
+//; eval_flags(c-sycl_cuda): [/usr/local/dpcpp-cuda/bin/clang++] -O3 -ldl -I/usr/local/dpcpp-cuda/include/sycl -L/usr/local/dpcpp-cuda/lib -lsycl -fsycl -fsycl-targets=nvptx64-nvidia-cuda-sycldevice -fsycl-unnamed-lambda -lpthread -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_75
```
Comment (ghostplant) [inline backends/c-sycl_intel/include/backend.hpp:6]: It doesn't require `--cuda-gpu-arch=sm_75` here since this compiling flag will only be use for codes that is not related to SYCL device codes. So you can safely remove it.

File: backends/c-sycl_intel/include/backend.hpp [hunk 2]
```diff
@@ -78,7 +75,7 @@ namespace ab {
     if (__BACKEND__ == "c-sycl_intel")
       ab_utils::Process({"dpcpp", path, "-std=c++17", "-lpthread", "-fPIC", "-shared", "-Wno-pass-failed", "-O3", "-ffast-math", "-march=skylake-avx512", "-o", path + ".out"}, 10);
     else
-      ab_utils::Process({"/usr/local/dpcpp-cuda/bin/clang++", path, "-std=c++17", "-ldl", "-fPIC", "-shared", "-O2", "-I/usr/local/dpcpp-cuda/include/sycl", "-L/usr/local/dpcpp-cuda/lib", "-lsycl", "-fsycl", "-fsycl-targets=nvptx64-nvidia-cuda-sycldevice", "-fsycl-unnamed-lambda", "-Wno-unknown-cuda-version", "-o", path + ".out"}, 20);
+      ab_utils::Process({"/usr/local/dpcpp-cuda/bin/clang++", path, "-std=c++17", "-ldl", "-fPIC", "-shared", "-O2", "-I/usr/local/dpcpp-cuda/include/sycl", "-L/usr/local/dpcpp-cuda/lib", "-lsycl", "-fsycl", "-fsycl-targets=nvptx64-nvidia-cuda-sycldevice", "-fsycl-unnamed-lambda", "-Wno-unknown-cuda-version", "-Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_75", "-o", path + ".out"}, 20);
```
Comment (ghostplant) [inline backends/c-sycl_intel/include/backend.hpp]: Different from command above, the compiling command here will target and only target to SYCL device code. Is it possible for SYCL-CUDA to query what kind of compute capability a certain CUDA GPU is? I think hard-coded `--cuda-gpu-arch=sm_75` will only work for specific GPU type.
For CUDA framework, there is [API](https://github.com/microsoft/antares/blob/v0.2.x/backends/c-rocm/include/backend.hpp#L81-L85) to query that. Is it possible for SYCL-CUDA to get that value?

File: backends/c-sycl_intel/include/backend.hpp (hunk 1 above)
Comment (Michoumichmich) [reply backends/c-sycl_intel/include/backend.hpp:6]: Oh okay I see, thanks, I'll remove that

File: backends/c-sycl_intel/include/backend.hpp (hunk 2 above)
Comment (Michoumichmich) [reply backends/c-sycl_intel/include/backend.hpp]: I'm not aware of a SYCL query that would do something similar. The issue is that without the backend version, the code is compiled for sm_50

Comment (ghostplant) [reply backends/c-sycl_intel/include/backend.hpp]: DPCPP for CUDA will always need CUDA Toolkit as one of dependencies right?

Comment (Michoumichmich) [reply backends/c-sycl_intel/include/backend.hpp]: Yes, the backend is built on top of that. So we can expect to have nvcc available and just use a script to get the compute capability.

I think we should be able to add the compute capability in the format `sm_XX` into the device descriptor `info::device::extensions`. The SYCL specification states that: *The extension names returned can be vendor supported extension names*. If so I could open a PR on DPC++ and return the compute capability through that query using  `CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR/MINOR`.

Comment (ghostplant) [reply backends/c-sycl_intel/include/backend.hpp]: That's fine if you want to extend SYCL properties.
On the other hand, because SYCL CUDA is always bind with CUDA Toolkit, another way which is much easier, is to append the link of `-lcuda` in the top `eval_flags` command, this make calling native CUDA API in the function `moduleLoad` possible, so you can also get that compatibility value for constructing the device code compiling arguments in `moduleLoad`.

Comment (Michoumichmich) [reply backends/c-sycl_intel/include/backend.hpp]: I see, that's interesting, I will do that while I see if we can solve that on the SYCL side.

Comment (Michoumichmich) [reply backends/c-sycl_intel/include/backend.hpp]: Hello, could you check that it works on your side? Including the CUDA headers in the same translation unit as SYCL was causing conflicts, I avoided the issue using `-iquote`.

Comment (ghostplant) [reply backends/c-sycl_intel/include/backend.hpp]: I get the following error using CUDA 10.0:
```
[EvalAgent] Compiling Evaluator: /usr/local/dpcpp-cuda/bin/clang++ /antares/backends/c-sycl_cuda/../../graph_evaluator/run_graph.cpp -D__BACKEND__=\"c-sycl_cuda\" -D__BACKEND_sycl_cuda__ -I/antares/backends/c-sycl_cuda/include -std=c++17 -Wno-string-compare -Wno-unused-result -Wno-unused-value -o /antares/.libAntares/evaluator.c-sycl_cuda.tmp -ldl -I/usr/local/dpcpp-cuda/include/sycl -L/usr/local/dpcpp-cuda/lib -lsycl -fsycl -fsycl-targets=nvptx64-nvidia-cuda-sycldevice -fsycl-unnamed-lambda -lpthread -iquote/usr/local/cuda/include -L/usr/local/cuda/lib64 -lcuda -DSYCL_CUDA
clang-11: warning: Unknown CUDA version. version.txt: 10.2.89. Assuming the latest supported version 10.1 [-Wunknown-cuda-version]
In file included from /antares/backends/c-sycl_cuda/../../graph_evaluator/run_graph.cpp:4:
In file included from /antares/backends/c-sycl_cuda/../../graph_evaluator/execute_module.hpp:80:
In file included from /antares/backends/c-sycl_cuda/include/backend.hpp:80:
/usr/local/cuda/include/cuda.h:240:8: error: expected unqualified-id
extern "C" {
       ^
In file included from /antares/backends/c-sycl_cuda/../../graph_evaluator/run_graph.cpp:4:
In file included from /antares/backends/c-sycl_cuda/../../graph_evaluator/execute_module.hpp:80:
/antares/backends/c-sycl_cuda/include/backend.hpp:82:48: error: use of undeclared identifier 'CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR'
    CHECK_OK(0 == cuDeviceGetAttribute(&major, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, 0));
                                               ^
/antares/backends/c-sycl_cuda/include/backend.hpp:83:48: error: use of undeclared identifier 'CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MINOR'
    CHECK_OK(0 == cuDeviceGetAttribute(&minor, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MINOR, 0));
                                               ^
3 errors generated.
```

Comment (ghostplant) [reply backends/c-sycl_intel/include/backend.hpp]: Is it possible that dpcpp automatically gets the device compute capability in runtime if not specifying any `--cuda-gpu-arch`?

Comment (Michoumichmich) [reply backends/c-sycl_intel/include/backend.hpp]: > I get the following error using CUDA 10.0:
> 
> ```
> [EvalAgent] Compiling Evaluator: /usr/local/dpcpp-cuda/bin/clang++ /antares/backends/c-sycl_cuda/../../graph_evaluator/run_graph.cpp -D__BACKEND__=\"c-sycl_cuda\" -D__BACKEND_sycl_cuda__ -I/antares/backends/c-sycl_cuda/include -std=c++17 -Wno-string-compare -Wno-unused-result -Wno-unused-value -o /antares/.libAntares/evaluator.c-sycl_cuda.tmp -ldl -I/usr/local/dpcpp-cuda/include/sycl -L/usr/local/dpcpp-cuda/lib -lsycl -fsycl -fsycl-targets=nvptx64-nvidia-cuda-sycldevice -fsycl-unnamed-lambda -lpthread -iquote/usr/local/cuda/include -L/usr/local/cuda/lib64 -lcuda -DSYCL_CUDA
> clang-11: warning: Unknown CUDA version. version.txt: 10.2.89. Assuming the latest supported version 10.1 [-Wunknown-cuda-version]
> In file included from /antares/backends/c-sycl_cuda/../../graph_evaluator/run_graph.cpp:4:
> In file included from /antares/backends/c-sycl_cuda/../../graph_evaluator/execute_module.hpp:80:
> In file included from /antares/backends/c-sycl_cuda/include/backend.hpp:80:
> /usr/local/cuda/include/cuda.h:240:8: error: expected unqualified-id
> extern "C" {
>        ^
> In file included from /antares/backends/c-sycl_cuda/../../graph_evaluator/run_graph.cpp:4:
> In file included from /antares/backends/c-sycl_cuda/../../graph_evaluator/execute_module.hpp:80:
> /antares/backends/c-sycl_cuda/include/backend.hpp:82:48: error: use of undeclared identifier 'CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR'
>     CHECK_OK(0 == cuDeviceGetAttribute(&major, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, 0));
>                                                ^
> /antares/backends/c-sycl_cuda/include/backend.hpp:83:48: error: use of undeclared identifier 'CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MINOR'
>     CHECK_OK(0 == cuDeviceGetAttribute(&minor, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MINOR, 0));
>                                                ^
> 3 errors generated.
> ```

It should be fixed

Comment (Michoumichmich) [reply backends/c-sycl_intel/include/backend.hpp]: > Is it possible that dpcpp automatically gets the device compute capability in runtime if not specifying any `--cuda-gpu-arch`?

It acts like `-march/-mtune`, if you don't specify the arch, if compiles to the lowest common denominator. But more specifically, `--cuda-gpu-arch` is a LLVM flag, not a DPC++ one. Maybe we could see if something like `--cuda-gpu-arch=native` would be interesting

File: backends/c-sycl_intel/include/backend.hpp [hunk 3]
```diff
@@ -77,9 +77,16 @@ namespace ab {
...
+    else {
+    std::string gpu_arch = "50"; // Corresponds to the back-end default.
+#ifdef SYCL_CUDA
+    int major, minor;
+    CHECK_OK(0 == cuDeviceGetAttribute(&major, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, 0));
+    CHECK_OK(0 == cuDeviceGetAttribute(&minor, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MINOR, 0));
+    gpu_arch = std::to_string(major * 10 + minor);
+#endif
+      ab_utils::Process({"/usr/local/dpcpp-cuda/bin/clang++", path, "-std=c++17", "-ldl", "-fPIC", "-shared", "-O2", "-I/usr/local/dpcpp-cuda/include/sycl", "-L/usr/local/dpcpp-cuda/lib", "-lsycl", "-fsycl", "-fsycl-targets=nvptx64-nvidia-cuda-sycldevice", "-fsycl-unnamed-lambda", "-Wno-unknown-cuda-version", "-Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_" + gpu_arch, "-o", path + ".out"}, 20);  
+    }
```
Comment (ghostplant) [inline backends/c-sycl_intel/include/backend.hpp:89]: Can you trim several indents of these code? I also found some lines end with multiple blanks.

File: backends/c-sycl_intel/include/backend.hpp (hunk 2 above)
Comment (ghostplant) [reply backends/c-sycl_intel/include/backend.hpp]: Thanks. It has been verified. Can you also fix some indents of those codes?

File: backends/c-sycl_intel/include/backend.hpp [hunk 4]
```diff
@@ -2,12 +2,15 @@
...
-//; eval_flags(c-sycl_cuda): [/usr/local/dpcpp-cuda/bin/clang++] -ldl -I/usr/local/dpcpp-cuda/include/sycl -L/usr/local/dpcpp-cuda/lib -lsycl -fsycl -fsycl-targets=nvptx64-nvidia-cuda-sycldevice -fsycl-unnamed-lambda -lpthread
+//; eval_flags(c-sycl_cuda): [/usr/local/dpcpp-cuda/bin/clang++] -ldl -I/usr/local/dpcpp-cuda/include/sycl -L/usr/local/dpcpp-cuda/lib -lsycl -fsycl -fsycl-targets=nvptx64-nvidia-cuda-sycldevice -fsycl-unnamed-lambda -lpthread -iquote/usr/local/cuda/include -L/usr/local/cuda/lib64 -lcuda -DSYCL_CUDA
 
 #include <CL/sycl.hpp>
 #include <dlfcn.h>
 #include <pthread.h>
 #include <malloc.h>
+#ifdef SYCL_CUDA
+#include "cuda.h"
+#endif
```
Comment (ghostplant) [inline backends/c-sycl_intel/include/backend.hpp:13]: Besides, can you make it from `"cuda.h"` to `<cuda.h>`?

Comment (Michoumichmich) [reply backends/c-sycl_intel/include/backend.hpp:13]: That is the issue I talked about before, if I move to `<cuda.h>`, I have to change the include from `-iquote` to `-I`, but then dpc++ picks up on the CUDA headers and I end up with a lot of errors such as:
```
In file included from /home/michel/sycl_workspace/deploy/include/sycl/CL/sycl/exception.hpp:15:
/home/michel/sycl_workspace/deploy/include/sycl/CL/sycl/detail/pi.h:251:7: error: use of undeclared identifier 'CL_DEVICE_QUEUE_ON_DEVICE_PROPERTIES'
      CL_DEVICE_QUEUE_ON_DEVICE_PROPERTIES,
      ^
/home/michel/sycl_workspace/deploy/include/sycl/CL/sycl/detail/pi.h:252:45: error: use of undeclared identifier 'CL_DEVICE_QUEUE_ON_HOST_PROPERTIES'
  PI_DEVICE_INFO_QUEUE_ON_HOST_PROPERTIES = CL_DEVICE_QUEUE_ON_HOST_PROPERTIES,
                                            ^
/home/michel/sycl_workspace/deploy/include/sycl/CL/sycl/detail/pi.h:256:31: error: use of undeclared identifier 'CL_DEVICE_IL_VERSION_KHR'
  PI_DEVICE_INFO_IL_VERSION = CL_DEVICE_IL_VERSION_KHR,
                              ^
/home/michel/sycl_workspace/deploy/include/sycl/CL/sycl/detail/pi.h:274:39: error: use of undeclared identifier 'CL_DEVICE_MAX_NUM_SUB_GROUPS'
  PI_DEVICE_INFO_MAX_NUM_SUB_GROUPS = CL_DEVICE_MAX_NUM_SUB_GROUPS,
                                      ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

Comment (ghostplant) [reply backends/c-sycl_intel/include/backend.hpp:13]: OK, it makes sense.

**Comment by ghostplant** [comment]
Great. Thanks for your PR @Michoumichmich !
